### ISSUE TITLE
Keycloak plugin does not work in RH SSO Openshift image

### ISSUE THREAD

**Issue by matskiv** [EnMasseProject/enmasse#3551]
**Describe the bug**
I added sasl plugin to a RH SSO image (downstream for Keycloak) but the ProtonServer throws an exception on startup - `io.vertx.core.VertxException: Key/certificate is mandatory for SSL` ([full stack trace](https://gist.github.com/matskiv/076484940d6bc70afc60a4ea2fe7733d)).

This seems to happen because of a different server config - [diff here](https://www.diffchecker.com/A61SIzsI).
Plugin seems to rely on a [configuration options](https://github.com/EnMasseProject/enmasse/blob/master/keycloak-plugin/src/main/configuration/standalone-h2.xml#L557-L569) which are not present [in configuration file for keycloak/RH SSO on openshift](https://github.com/jboss-container-images/redhat-sso-7-openshift-image/blob/sso73-dev/modules/sso/config/launch/setup/73/added/standalone-openshift.xml).

**To Reproduce**
Deploy RH SSO with sasl plugin(.jar I used is hosted [here](https://github.com/matskiv/sasl-plugin-jar/blob/master/sasl-plugin-0.30.0.jar)) added to `/opt/eap/providers`
Observe keycloak pod logs to see an exception mentioned above.

**Expected behavior**
No exceptions in keycloak pod logs.
Plugin working as expected.

**Additional context**
I am working on Integreatly project, we would like to use external Keycloak instance because we already deploy one, and we have better control of it (HA configuration, monitoring, etc.).

**Comment by lulf** [comment]
@matskiv It looks like it is missing a certificate to be configured. Currently the plugin relies on this configuration:

```
<spi name="amqp">
  <provider name="amqp-server" enabled="true">
    <properties>
      <property name="enableTls" value="true"/>
      <property name="enableNonTls" value="false"/>
      <property name="pemKeyPath" value="/opt/enmasse/cert/tls.key"/>
      <property name="pemCertificatePath" value="/opt/enmasse/cert/tls.crt"/>
      <property name="tlsHost" value="0.0.0.0"/>
      <property name="tlsPort" value="5671"/>
      <property name="defaultDomain" value="enmasse"/>
    </properties>
  </provider>
</spi>
```

Are the pemKeyPath and pemCertififcatPath present in your deployment?

**Comment by matskiv** [comment]
@lulf The pemKeyPath and pemCertificatePath are not in the configuration that is used by RH SSO.
You can see it here btw: https://github.com/jboss-container-images/redhat-sso-7-openshift-image/blob/sso73-dev/modules/sso/config/launch/setup/73/added/standalone-openshift.xml

And just to add more context here, I am using [keycloak operator](https://github.com/keycloak/keycloak-operator), so I can't really edit that configuration.

The plugin seems to check few different config variables when configuring ssl [here](https://github.com/EnMasseProject/enmasse/blob/master/keycloak-plugin/sasl-plugin/src/main/java/io/enmasse/keycloak/spi/AmqpServer.java#L162-L179). Maybe it would be possible to add another branch to that code, which would work with configuration used in RH SSO?

**Comment by lulf** [comment]
@matskiv Yes, the plugin is not part of keycloak, so it is not in the standard keycloak configuration. If this configuration cannot be edited, I don't see how you can enable the plugin. IMO the operator  need to support some way of 'injecting' the config needed by the plugin, otherwise it cannot be enabled.

**Comment by matskiv** [comment]
@lulf I have very little experience with EAP and configuration of the apps, so I apologize in advance if I am talking nonsense.

SSL configuration IMHO isn't enmasse specific, it's used by the "host" Keycloak as well. So why can't sasl plugin pull SSL config from the same place as Keycloak/RHSSO does? Is that technically not possible?

P.S: If the plugin code would be able to use the default config, maybe you even wouldn't have to maintain config files here - ` enmasse/keycloak-plugin/src/main/configuration/standalone-*.xml`.

**Comment by lulf** [comment]
@matskiv No worries, i have very little experience with it as well :) 

Regarding certs, I agree, the plugin code could be changed to use the default Keystore instead.

**Comment by matskiv** [comment]
@lulf The existing configuration properties (enableTls, enableNonTls, pemKeyPath, pemCertificatePath, tlsHost, tlsPort, defaultDomain) could be kept as optional overrides.
And if the overrides are not provided, plugin would use configuration of "host" keycloak(e.g. host, SSL certs) or reasonable default value (e.g. port).

**Comment by k-wall** [comment]
AFAIK keycloak-plugin at the moment is really just an internal concern within EnMasse that is released in lockstep with EnMasse itself.  The use-case you suggest essentially promotes it to be a component with a public API  I think to support this use-case it would need some work, such as:

1. supporting different EnMasse instance / privacy 
2. exposing metrics for monitoring/diagnostic purposes
3. API versioning (so backward compatibility could be retained if the API needs to be changed).

Also the component at the moment is not published independently too.

**Comment by matskiv** [comment]
A support for external Keycloak instance was added to EnMasse in https://github.com/EnMasseProject/enmasse/issues/682, so based on that, the plugin isn't necessarily just an internal concern anymore (IMHO). And with this issue I am simply asking to make it possible to use the plugin with external Keycloak without having to inject custom configuration (for the default config).
Unfortunately I am lacking knowledge of underlying technologies (Vert.x/EAP/Java) to implement the necessary changes myself, but I would be happy to provide you a pre-configured testing environment for this use case and help with testing.

As for the points from previous comment:
1. Do you mean multiple EnMasse instances? Why is it necessary to support this?
2. What auth related metrics does EnMasse currently expose? Would the plugin need code changes to expose same metrics when used in external Keycloak?
3. Backwards compatibility could be optional. Updating sasl plugin and keycloak in lockstep with EnMasse is IMHO reasonable prerequisite for the support of external Keycloak.

**Comment by maleck13** [comment]
@k-wall @lulf in our use case we have an Keycloak setup that is resilient and makes use of a HA database setup. If we want to use the feature to use external Keycloak, what is the recommend way to manage users and authentication? Isn't the plugin needed for that feature to work?
Alternatively, if the above is just not possible at the moment, is there a way to configure EnMasse to scale up the keycloak provisioned by the standard authservice and have it connect to an external database ?

**Comment by lulf** [comment]
@matskiv @maleck13 @k-wall 

> A support for external Keycloak instance was added to EnMasse in #682, so based on that, the plugin isn't necessarily just an internal concern anymore (IMHO). And with this issue I am simply asking to make it possible to use the plugin with external Keycloak without having to inject custom configuration (for the default config).

I think it makes sense to make the sasl-plugin work with a generic Keycloak, and ideally the keycloak-operator, I totally agree on that. In turn, that would also benefit us as we could then also use the keycloak-operator to deploy the builtin standard authentication service.

> In our use case we have an Keycloak setup that is resilient and makes use of a HA database setup. If we want to use the feature to use external Keycloak, what is the recommend way to manage users and authentication? Isn't the plugin needed for that feature to work?
Alternatively, if the above is just not possible at the moment, is there a way to configure EnMasse to scale up the keycloak provisioned by the standard authservice and have it connect to an external database ?

If you want to use a Keycloak instance external to EnMasse, then you want to configure an external authenticationservice and point that to your Keycloak instance. That instance would then need to contain the plugin. 

An alternative to that is to configure a standard authentication service and point it to use an external database as the backend. Currently, we support using a postgresql database as the data source (see https://enmasse.io/documentation/0.30.1/openshift/#assembly-auth-services-messaging). We do not support deploying HA instances of the keycloak instance itself at this point.

If you go down the route of using an external Keycloak, you will need to manage users and permissions in Keycloak yourself, this is outside the scope of EnMasse. The MessagingUser CRD is only synced with the standard authservice that we control. Now that the MessagingUser CR is actually a CRD, this means you can write your own controller to sync users to the external authservice if you wish. 

That being said, now that users are native CRDs (see #3535), the source of truth of MessagingUsers are actually stored in etcd.  At this point, the keycloak instance in the standard authservice is just used as a lookup mechanism for users originally stored in CRDs, so I wonder if we should replace the standard authservice with something more lightweight than Keycloak.

**Comment by matskiv** [comment]
@lulf I was referencing this issue https://github.com/EnMasseProject/enmasse/issues/682 when talking about "external Keycloack". If I understood it correctly, we would be able to use "standard" AuthenticationService and external Keycloak instance if we override "keycloak-credentials" in enmasse namespace. If we did that, user and realm managment would be done on our Keycloak instance, instead of built-in one, right?

**Comment by lulf** [comment]
@matskiv That capability is no longer part of and supported in enmasse. Earlier when keycloak-controller was a separate deployment, it was possible to point that component to a separate keycloak instance. However, this feature was never used and we've introduced the AuthenticationService CRD since then, which defines a boundary between standard authservice, which is happens to be an instance of Keycloak at the time being, and an external authentication service, which is a keycloak not managed by enmasse.

**Comment by matskiv** [comment]
@lulf I see. Thanks for clarifying.

In our use case the external keycloak will be shared. 
Also, users of that keycloak will be able to create realms with arbitrary name, so name collision with enmasse is a concern.

I like the idea of a lightweight authservice that would just get/watch MessagingUsers resources.
It can be implemented as an external AuthenticationService first, and if you like, it can replace standard AuthenticationService (and built-in Keycloack). I'll give it a shot in spare time.

**Comment by lulf** [comment]
Closing this, as we will not be able to fix this for the 0.32 release, and the release after that will not have Keycloak installed. (See #4391 )
